### ISSUE TITLE
Accessing parameters and form variables in request guard

### ISSUE THREAD

**Issue by InstinctBas** [SergioBenitez/Rocket#914]
I might be going down the wrong path here, still pretty new to rocket.

I'm trying to implement a request guard that handles a part of the authentication that is the same for most of my routes. The idea is that it checks for an authentication token, checks if that token is still valid and if so, returns that token (and if not it fails).

The problem that I have is that there are multiple methods for sending the token to my web server depending on the abilities of whomever is accessing the server:
- preferred way is in an authentication header (this bit I have working and is easy)
- as an optional parameter in the URL
- as a hidden form variable
- as a cookie

I already read elsewhere that we don't have access to the request body in a request guard so that might rule out the ability to handle a form variable, I guess I can live with that.

As an optional parameter I seem to be able to call get_param on the request but it takes an index, I guess I could loop it to see if I have a parameter called token?

I haven't looked into cookies yet :)

Using rocket 0.4.0

P.S. is there a discord/IRC/matrix/slack channel around for Rocket by any chance?

**Comment by jebrosen** [comment]
> The problem that I have is that there are multiple methods for sending the token to my web server depending on the abilities of whomever is accessing the server:

This high a level of flexibility in a request isn't made easy by the current API surface, unfortunately. Here's what I think of the specific options you mentioned:

>    * preferred way is in an authentication header (this bit I have working and is easy)

Via `FromRequest`, I assume? It's good to hear you got that working easily!

>    * as an optional parameter in the URL

This is trickier. Path parameters (`FromParam`) don't have access to managed state (see also #404). I think currently your options are to loop through them in `FromRequest` (which is ugly, and you have no way to know which parameter is which).

Query parameters should be accessible via `get_query_value`, though, which can be done in a type that implements `FromRequest`.

>    * as a hidden form variable
>
> I already read elsewhere that we don't have access to the request body in a request guard so that might rule out the ability to handle a form variable, I guess I can live with that.

Yes, this is currently impossible except via a data guard, and even then you have to either assume it's in the peek buffer available on `Data` or consume the entire body in the process of finding it. Issue #775 might make this a bit easier to deal with.

>    * as a cookie

Cookies are also available in `FromRequest`. Watch out for the "multiple active `Cookies` instance" error if you use your validation request guard and request cookies in the same route.

In summary, I think with a single `FromRequest` implementation you can get to a header, a (consistently named) query string parameter, and a cookie. That leaves out form fields and path parameters, which might have to be done in the actual routes by reading the form values or by checking a `FromParam` value if you want them to be checkable.

> P.S. is there a discord/IRC/matrix/slack channel around for Rocket by any chance?

There are a few resources listed at https://github.com/SergioBenitez/Rocket#documentation -- including an IRC channel also bridged to matrix.

**Comment by jebrosen** [comment]
Closing because it looks like this has been answered; please re-open if I missed something.

**Comment by InstinctBas** [comment]
Thanks Jeb,

Sorry, yes I managed to get as far as I needed to go, got this working with authentication, get parameters and cookies. Left post alone ;)
